vyper/venom/analysis/available_expression.py

```diff
@@ -0,0 +1,257 @@
+        if self.is_commutative:
+            if self.operands[0].same(other.operands[1]) and self.operands[1].same(
+                other.operands[0]
+            ):
+                return True
+
+        # General case
+        for self_op, other_op in zip(self.operands, other.operands):
+            if not self_op.same(other_op):
+                return False
+
+        return True
+
+    def __hash__(self) -> int:
+        return hash(self.first_inst)
+
+    def __repr__(self) -> str:
+        if self.opcode == "store":
+            assert len(self.operands) == 1, "wrong store"
+            return repr(self.operands[0])
+        res = self.opcode + " [ "
+        for op in self.operands:
+            res += repr(op) + " "
+        res += "]"
+        return res
+
+    @cached_property
+    def get_depth(self) -> int:
+        max_depth = 0
+        for op in self.operands:
+            if isinstance(op, _Expression):
+                d = op.get_depth
+                if d > max_depth:
+                    max_depth = d
+        return max_depth + 1
+
+    @cached_property
+    def get_reads(self) -> Effects:
+        tmp_reads = self.first_inst.get_read_effects()
+        for op in self.operands:
+            if isinstance(op, _Expression):
+                tmp_reads = tmp_reads | op.get_reads
+        if self.ignore_msize:
+            tmp_reads &= ~Effects.MSIZE
+        return tmp_reads
+
+    @cached_property
+    def get_writes(self) -> Effects:
+        tmp_reads = self.first_inst.get_write_effects()
+        for op in self.operands:
+            if isinstance(op, _Expression):
+                tmp_reads = tmp_reads | op.get_writes
+        if self.ignore_msize:
+            tmp_reads &= ~Effects.MSIZE
+        return tmp_reads
+
+    @property
+    def is_commutative(self) -> bool:
+        return self.first_inst.is_commutative
+
+
+class AvailableExpressionAnalysis(IRAnalysis):
+    inst_to_expr: dict[IRInstruction, _Expression]
+    dfg: DFGAnalysis
+    inst_to_available: dict[IRInstruction, OrderedSet[_Expression]]
+    bb_outs: dict[IRBasicBlock, OrderedSet[_Expression]]
+
+    # the size of the expressions
+    # that are considered in the analysis
+    min_depth: int
+    max_depth: int
+    ignore_msize: bool
+
+    def __init__(
+        self,
+        analyses_cache: IRAnalysesCache,
+        function: IRFunction,
+        min_depth: int = _MIN_DEPTH,
+        max_depth: int = _MAX_DEPTH,
+    ):
+        super().__init__(analyses_cache, function)
+        self.analyses_cache.request_analysis(CFGAnalysis)
+        dfg = self.analyses_cache.request_analysis(DFGAnalysis)
+        assert isinstance(dfg, DFGAnalysis)
+        self.dfg = dfg
+
+        self.min_depth = min_depth
+        self.max_depth = max_depth
+
+        self.inst_to_expr = dict()
+        self.inst_to_available = dict()
+        self.bb_outs = dict()
+
+        self.ignore_msize = not self._contains_msize()
```

Comment:
maybe we should generalize this. basically what this is saying is, "we don't need to care about making memory reads volatile when there is no msize". but it's not that clear from how the code is written. (e.g. maybe there are other write effects which we don't need to preserve if there is no subsequent read).